reference-implementation/lib/readable-stream.js

```diff
@@ -1935,3 +1934,16 @@ function byteStreamControllerBrandCheckException(name) {
   return new TypeError(
     `ReadableByteStreamController.prototype.${name} can only be used on a ReadableByteStreamController`);
 }
+
+// Helper function for ReadableStream pipeThrough
+
+function hasPromiseIsHandledInternalSlot(promise) {
+  try {
+    // This relies on the brand-check that is enforced by Promise.prototype.then(). As with the rest of the reference
+    // implementation, it doesn't attempt to do the right thing if someone has modified the global environment.
+    Promise.prototype.then.call(promise, undefined, () => {});
```

Comment:
This is not side-effect free; it marks the promise as handled itself. Calling `then(undefined, undefined)` somehow prevents the promise from subsequently being marked as handled. I don't know whether that is standard behaviour or a bug in V8.

In the specific case where we are using it, this doesn't matter, because when this function runs true we always mark the promise as handled anyway.